### ISSUE TITLE
'unique' consumes sequences

### ISSUE THREAD

**Issue by moon-chilled** [rakudo/rakudo#4513]
```
$ raku -e "say [[1], ['1'], [4]].unique(:as(*.map(&[~])), :with(&[eqv]))"
The iterator of this Seq is already in use/consumed by another Seq (you
might solve this by adding .cache on usages of the Seq, or by assigning
the Seq into an array)
  in block <unit> at -e line 1
```

Changing the key to `*.map(&[~]).cache` works around the issue

**Comment by jubilatious1** [comment]
```
> say [[1], ['1'], [4]].unique(:as(*.map(&[~])), :with(&[eqv]))
The iterator of this Seq is already in use/consumed by another Seq
(you might solve this by adding .cache on usages of the Seq, or
by assigning the Seq into an array)
  in block <unit> at <unknown file> line 1

> say [[1], ['1'], [4]];
[[1] [1] [4]]
> say [[1], ['1'], [4]].WHAT
(Array)
> my @array = [[1], ['1'], [4]]
[[1] [1] [4]]
> @array.unique(:as(*.map(&[~])), :with(&[eqv]))
The iterator of this Seq is already in use/consumed by another Seq
(you might solve this by adding .cache on usages of the Seq, or
by assigning the Seq into an array)

> $*VM
moar (2021.06)
```

**Comment by lizmat** [comment]
Test added with https://github.com/Raku/roast/commit/8c0928daf4

**Comment by jubilatious1** [comment]
```perl6
 % raku
Welcome to Rakudo™ v2023.05.
Implementing the Raku® Programming Language v6.d.
Built on MoarVM version 2023.05.

To exit type 'exit' or '^D'
[0] > say [[1], ['1'], [4]].unique(:as(*.map(&[~])), :with(&[eqv]))
The iterator of this Seq is already in use/consumed by another Seq (you
might solve this by adding .cache on usages of the Seq, or by assigning
the Seq into an array)
  in block <unit> at <unknown file> line 1

[0] > say [[1], ['1'], [4]].List.unique(:as(*.map(&[~])), :with(&[eqv]))
The iterator of this Seq is already in use/consumed by another Seq (you
might solve this by adding .cache on usages of the Seq, or by assigning
the Seq into an array)
  in block <unit> at <unknown file> line 1

[0] > say [[1], ['1'], [4]].unique
([1] [1] [4])
[0] > say [[1], ['1'], [4]].unique(:as(*.Str))
([1] [4])
[0] > say [[1], ['1'], [4]].unique(:as(~*))
([1] [4])
[0] >
```
@moon-chilled

**Comment by lizmat** [comment]
It's **not** fixed in 2023.05.  It is fixed in HEAD now, and will be in the 2024.10 release.

**Comment by jubilatious1** [comment]
> It's **not** fixed in 2023.05. It is fixed in HEAD now, and will be in the 2024.10 release.

Pointing out other existing workarounds.
